src/lib/es2015.core.d.ts

```diff
@@ -20,50 +30,50 @@ interface Array<T> {
      * @param thisArg If provided, it will be used as the this value for each invocation of
      * predicate. If it is not provided, undefined is used instead.
      */
-    findIndex(predicate: (value: T, index: number, obj: T[]) => unknown, thisArg?: any): number;
+    findIndex(predicate: (value: T, index: number, array: T[]) => unknown, thisArg?: any): number;
 
     /**
      * Changes all array elements from `start` to `end` index to a static `value` and returns the modified array
      * @param value value to fill array section with
      * @param start index to start filling the array at. If start is negative, it is treated as
-     * length+start where length is the length of the array.
+     * length + start where length is the length of the array.
      * @param end index to stop filling the array at. If end is negative, it is treated as
-     * length+end.
+     * length + end.
      */
-    fill(value: T, start?: number, end?: number): this;
+    fill(value: T, start?: number, end?: number): T[];
 
     /**
      * Returns the this object after copying a section of the array identified by start and end
      * to the same array starting at position target
-     * @param target If target is negative, it is treated as length+target where length is the
+     * @param target If target is negative, it is treated as length + target where length is the
      * length of the array.
-     * @param start If start is negative, it is treated as length+start. If end is negative, it
-     * is treated as length+end.
+     * @param start If start is negative, it is treated as length + start. If end is negative, it
+     * is treated as length + end.
      * @param end If not specified, length of the this object is used as its default value.
      */
-    copyWithin(target: number, start: number, end?: number): this;
+    copyWithin(target: number, start: number, end?: number): T[];

```

Comment:
possible break: subtypes of Array will now just be Array<T>, missing any of the subtype methods.

Is the only reason for this change consistency?

Reply:
This has to be long.

Returning `this` in `{Array, %TypedArray%}.prototype.{fill, copyWithin, sort}` *seems* to be correct, but TypeScript's type system is purely type-of-value-based, not instance-based. And because a subtype such as a tuple, or the generic parameters of a type, can contain the information of the values, returning `this`, which means an object with a same type, is technically incorrect for these methods which mutate the original object in a disruptive way. For example:

```ts
declare const foo: [42, 1, 2, 3];
const bar = foo.fill(42); // !!?

interface ArraySubtype extends Array<string> {
	0: "foo";
}
const bar: ArraySubtype = ["foo", "bar", "baz"];
const bar1 = bar.sort(); // !!!

interface Int8ArraySubtype extends Int8Array {
	0: 42;
}
const baz = new Int8Array([42, 1, 2, 3]) as Int8ArraySubtype;
const baz1 = baz.sort(); // ?!!
```

In fact, I'd [done the some thing](https://github.com/microsoft/TypeScript/pull/50452/files#diff-bc5657cc7554ca0c944ccc883a1334694829bce1293fd4fc5f2059cada6c003fR988) on `RegExp` (it's a deprecated method though):

https://github.com/microsoft/TypeScript/blob/facef94fc69d796d1d208585d7f3b0a8b7b0d00e/src/lib/es5.d.ts#L988

I understand these are all rare cases, but since they return the same instance, people who find the original type desirable could just use the original thing. It just makes sense, and just *because* these are all rare cases it has to be corrected.

---

In contrast, the following is better typed `this`. That's why the `this` keyword in type level is misleading:

```ts
class Foo<T> {
	constructor(public foo: T) {}
	clone(): this {
		return Object.create(Object.getPrototypeOf(this), Object.getOwnPropertyDescriptors(this));
	}
}
class Bar<T> extends Foo<T> {
	bar = true;
}
new Bar(42).clone(); // still a Bar<number>
```

Reply:
A use case of the above:

```ts
interface Array<T> {
	slice(): this;
}
```

(But then for `ReadonlyArray`, should the same method overload be typed `this & T[]`?)

(idea from https://github.com/microsoft/TypeScript/issues/36554#issuecomment-845292965)

Reply:
Is there a bug for this? Like the flatMap change in es2019, this isn't worth fixing unless it's breaking somebody's actual code, because any fix will break other code. Backward compatibility like this is the reason, for example, that we haven't been able to make overriding methods inherit parameter types.